### ISSUE TITLE
Issue spritebach and custom shaders (potentially other shader usage) [WinDX] (example code)

### ISSUE THREAD

**Issue by pfferrao** [mono/MonoGame#3700]
It appears to be an bug with setting a texture on sampler, if an rendertarget swap occurs.

```C#
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;

namespace DTBug {
    /// <summary>
    /// This is the main type for your game
    /// </summary>
    public class Game1 : Game {
        GraphicsDeviceManager graphics;
        SpriteBatch spriteBatch;
        Texture2D texture_0, texture_1;
        Vector2 sk_ratio, sk_jitter;
        RenderTarget2D rt;
        Effect effect;

        public Game1()
            : base() {
            graphics = new GraphicsDeviceManager(this);
            Content.RootDirectory = "Content";
        }

        /// <summary>
        /// Allows the game to perform any initialization it needs to before starting to run.
        /// This is where it can query for any required services and load any non-graphic
        /// related content.  Calling base.Initialize will enumerate through any components
        /// and initialize them as well.
        /// </summary>
        protected override void Initialize() {
            // TODO: Add your initialization logic here

            base.Initialize();
        }

        /// <summary>
        /// LoadContent will be called once per game and is the place to load
        /// all of your content.
        /// </summary>
        protected override void LoadContent() {
            // Create a new SpriteBatch, which can be used to draw textures.
            spriteBatch = new SpriteBatch(GraphicsDevice);

            // TODO: use this.Content to load your game content here
            texture_0 = Content.Load<Texture2D>(@"mono");
            texture_1 = Content.Load<Texture2D>(@"sketch");
            effect = Content.Load<Effect>(@"title");

            rt = new RenderTarget2D(GraphicsDevice, graphics.PreferredBackBufferWidth, graphics.PreferredBackBufferHeight, true, graphics.PreferredBackBufferFormat, graphics.PreferredDepthStencilFormat);

            sk_ratio = new Vector2(
                texture_0.Width / (float)texture_1.Width,
                texture_0.Height / (float)texture_1.Height);
        }

        /// <summary>
        /// UnloadContent will be called once per game and is the place to unload
        /// all content.
        /// </summary>
        protected override void UnloadContent() {
            // TODO: Unload any non ContentManager content here
        }

        /// <summary>
        /// Allows the game to run logic such as updating the world,
        /// checking for collisions, gathering input, and playing audio.
        /// </summary>
        /// <param name="gameTime">Provides a snapshot of timing values.</param>
        protected override void Update(GameTime gameTime) {
            if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed || Keyboard.GetState().IsKeyDown(Keys.Escape))
                Exit();

            // TODO: Add your update logic here

            base.Update(gameTime);
        }

        /// <summary>
        /// This is called when the game should draw itself.
        /// </summary>
        /// <param name="gameTime">Provides a snapshot of timing values.</param>
        protected override void Draw(GameTime gameTime) {

            GraphicsDevice.SetRenderTarget(rt);


            GraphicsDevice.SetRenderTarget(null);
            GraphicsDevice.Clear(Color.CornflowerBlue);
            
            effect.Parameters["sk_ratio"].SetValue(sk_ratio);
            effect.Parameters["sk_jitter"].SetValue(sk_jitter);
            effect.Parameters["sk_texture"].SetValue(texture_1);
            effect.CurrentTechnique = effect.Techniques[0];

            spriteBatch.Begin(0, null, null, null, null, effect);
            spriteBatch.Draw(texture_0, new Vector2(120, 120), Color.White);
            spriteBatch.End();

            base.Draw(gameTime);
        }
    }
}
```
The shader effect title.fx:
```GLSL
#include "Macros.fxh"

#ifdef SM4
#define TECHNIQUE_FRAGMENT_ONLY(name, psname) \
    technique name { pass { PixelShader = compile ps_4_0_level_9_1  psname(); } }
#else
#define TECHNIQUE_FRAGMENT_ONLY(name, psname ) \
    technique name { pass { PixelShader = compile ps_2_0  psname(); } }
#endif

DECLARE_TEXTURE(sb_texture, 0);
DECLARE_TEXTURE(sk_texture, 1);

BEGIN_CONSTANTS

    float2 sk_ratio;
    float2 sk_jitter;

END_CONSTANTS

float4 SketchPixelShader(float4 position : SV_Position, float4 color : COLOR0, float2 tex_coord : TEXCOORD0) : SV_Target0 {

    float4 color1 = SAMPLE_TEXTURE(sb_texture, tex_coord);
    float4 color2 = SAMPLE_TEXTURE(sk_texture, (tex_coord + sk_jitter) * sk_ratio);
    return color1 * color2;
}

TECHNIQUE_FRAGMENT_ONLY(SketchTechnique  , SketchPixelShader  );
```
Expected result (if the rendertarget isn't changed or by running on WinGL):
![expected](https://cloud.githubusercontent.com/assets/9132168/7010629/ac5496ce-dc9a-11e4-9c23-981b7e7f8fc0.png)

However if the rendertarget swap occurs the image will quickly flash and then nothing gets renderered. It appears that the current samplers state aren't getting properly apply.

Adding the following line after setting the shader parameters "fixes" the issue:
```C#
GraphicsDevice.Textures[1] = null;
```
Since the shaders values are passed to device inside the apply() during the spritebatch operation, the, it appears by setting texture in s1 to null we are setting some dirty flag that isn't properly set when the rendertarget swap occurs.

More information in the following link:
http://community.monogame.net/t/issue-with-spritebach-and-custom-shaders/2196

_Edit:_ adds more details and example code

**Comment by tgjones** [comment]
@tomspilman I think this is the same issue as #3710, so it can also be closed now that #3720 has been merged.

**Comment by tomspilman** [comment]
@pfferrao - Let us know if you still have issues with this.
